### PR TITLE
feat: Add Async refresh to Sql Registry

### PR DESCRIPTION
**Pull request by stanconia**
# What this PR does / why we need it:
Support Async refresh in Sql Registry to facilitate Registry functions at an Enterprise level

# Which issue(s) this PR fixes:
Registry Refresh at an enterprise scale

# Fixes

### CODE REVIEW

**Comment by stanconia** [comment]
ok-to-test

**Comment by tokoko** [comment]
I'm still not sure if I'm opposed to the idea, but just to note: don't we already do something like this "externally", for example a python feature server has an async methods that [refresh](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L76) registry in the background?

**Comment by stanconia** [comment]
> I'm still not sure if I'm opposed to the idea, but just to note: don't we already do something like this "externally", for example a python feature server has an async methods that [refresh](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L76) registry in the background?

Thanks for pointing this out, will evaluate and if good i will close pr.

**Comment by stanconia** [comment]
> I'm still not sure if I'm opposed to the idea, but just to note: don't we already do something like this "externally", for example a python feature server has an async methods that [refresh](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L76) registry in the background?



> I'm still not sure if I'm opposed to the idea, but just to note: don't we already do something like this "externally", for example a python feature server has an async methods that [refresh](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L76) registry in the background?

@tokoko problem with async refresh changes in Feature Server are, multiple feature services will call registry's refresh() hence overloading the system. We believe aync refresh should be done at the cacheRegistry level.

**Comment by stanconia** [comment]
> @stanconia I don't particularly buy that argument tbh. If you're running multiple feature servers, each will almost certainly be running in it's own process/container, so each will come with it's own `CachingRegistry`. The behavior will be exactly the same from a load perspective. Having said that, decoupling all that from feature server and making it more reusable might be a good idea. I'll have to take a closer look at the components involved first and get back to you (feature server, ui server, caching registry, ...)

You are right for some usecases but in our usecase we are using  a Remote Registry proxied to a Sql Registry.

**Comment by tokoko** [comment]
@stanconia oh, great. makes sense. didn't know anyone was already using it, I still have to add remote registry to the docs :)

**Comment by stanconia** [comment]
> @stanconia oh, great. makes sense. didn't know anyone was already using it, I still have to add remote registry to the docs :)

@tokoko gentle reminder on reviewing/considering this pr.

**Comment by tokoko** [comment]
How about instead of `allow_async_cache`, we introduce `cache_mode` config that can be `sync` (default), `thread` (implementation in this PR) and in the future possibly `async` (similar to this, but with asyncio). async referring to a separate thread in python is a little confusing at first.

**Comment by stanconia** [comment]
> How about instead of `allow_async_cache`, we introduce `cache_mode` config that can be `sync` (default), `thread` (implementation in this PR) and in the future possibly `async` (similar to this, but with asyncio). async referring to a separate thread in python is a little confusing at first.

@tokoko  requested change has been implemented.

File: sdk/python/feast/infra/registry/caching_registry.py
Comment (tokoko) [inline sdk/python/feast/infra/registry/caching_registry.py:315]: This will just run once, won't it? I think you meant to call refresh first and spawn a thread calling `_start_thread_async_refresh` to create an infinite loop. Also you shouldn't start a thread if `cache_ttl_seconds == 0`

Comment (stanconia) [reply sdk/python/feast/infra/registry/caching_registry.py:317]: > This will just run once, won't it? I think you meant to call refresh first and spawn a thread calling `_start_thread_async_refresh` to create an infinite loop. Also you shouldn't start a thread if `cache_ttl_seconds == 0`

@tokoko Threading.Timer will spin up a new thread and run a function at specified intervals hence it does not only run once. It is an infinite loop.  `cache_ttl_seconds > 0` implemented

Comment (tokoko) [reply sdk/python/feast/infra/registry/caching_registry.py:317]: Can you doublecheck? Just tried running this `threading.Timer(2, lambda: print('Hello World')).start()` and it only ran once. docs don't mention anything about Timer creating an infinite loop either. Take a look at how it's implemented in [feature server](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L77).

Comment (stanconia) [reply sdk/python/feast/infra/registry/caching_registry.py]: @tokoko you are correct, it is executed once. I made adjustments to run in an infinite loop.

File: sdk/python/feast/infra/registry/caching_registry.py
Comment (tokoko) [inline sdk/python/feast/infra/registry/caching_registry.py:323]: This shouldn't be called on each iteration. Maybe move it to the constructor and register only if `cache_ttl_seconds > 0`

File: sdk/python/feast/infra/registry/caching_registry.py
Comment (tokoko) [inline sdk/python/feast/infra/registry/caching_registry.py:317]: I think this should be called once even if cache_ttl_seconds isn't configured as infinite cache is implied

Comment (stanconia) [reply sdk/python/feast/infra/registry/caching_registry.py]: done.

File: sdk/python/feast/infra/registry/caching_registry.py
Comment (stanconia) [reply sdk/python/feast/infra/registry/caching_registry.py]: done

File: sdk/python/feast/repo_config.py
Comment (EXPEbdodla) [inline sdk/python/feast/repo_config.py:133]: Please write the docstring for this parameter. So all users know what modes are available.

File: sdk/python/feast/infra/registry/caching_registry.py
Comment (EXPEbdodla) [inline sdk/python/feast/infra/registry/caching_registry.py:319]: Assume this creates new threads for each call in recursion? Can you check and confirm once?

Comment (EXPEbdodla) [reply sdk/python/feast/infra/registry/caching_registry.py:319]: Sorry. Not able to reproduce this.

Comment (tokoko) [reply sdk/python/feast/infra/registry/caching_registry.py:319]: @EXPEbdodla The way I read it, each new thread will 1) refresh 2) spawn off another timer thread (creating an infinite loop) and exit. Am I missing something?

Comment (EXPEbdodla) [reply sdk/python/feast/infra/registry/caching_registry.py:319]: That's right. I was under impression we need to explicitly cancel/stop the spawned thread. Timer class is doing that automatically.

Comment (tokoko) [reply sdk/python/feast/infra/registry/caching_registry.py:319]: Yeah, this is also how it's currently implemented in python feature server. Probably not the best way though as each iteration creates a brand new os-level thread. We might have to revisit that later.

**Comment by tokoko** [comment]
@jeremyary the tests need approval here

**Comment by stanconia** [comment]
> > @jeremyary the tests need approval here
> 
> @jeremyary any updates? cc @tokoko

@tokoko can you assist here?

**Comment by tokoko** [comment]
@franciscojavierarceo @HaoXuAI can you help here? Tests need a `first-time contributor` approval.

**Comment by HaoXuAI** [comment]
hey, sorry merged amother PRs and caused conflicts. mind resolve it? I'll follow up from there.

**Comment by stanconia** [comment]
> hey, sorry merged amother PRs and caused conflicts. mind resolve it? I'll follow up from there.

@HaoXuAI resolved conflicts please approve

**Review by HaoXuAI — APPROVED** [review]
lgtm

**Comment by stanconia** [comment]
> lgtm

@tokoko  please review.

**Review by tokoko — APPROVED** [review]
lgtm

**Comment by stanconia** [comment]
> 2 workflows awaiting approval

@tokoko how do i get approvals for pending workflows?

**Comment by franciscojavierarceo** [comment]
@tokoko can you try /lgtm

**Comment by stanconia** [comment]
> /lgtm

@tokoko  please approve lint workflows

**Review by franciscojavierarceo — APPROVED** [review]

**Comment by stanconia** [comment]
> > /lgtm
> 
> @tokoko please approve lint workflows

@tokoko please can you assist with a merge

**Comment by tokoko** [comment]
@stanconia I don't have merge rights, Francisco can merge as well

**Comment by franciscojavierarceo** [comment]
Hi @stanconia I was going to give you a shoutout in the latest food newsletter if you wanted on. Feel free to add your name here: https://docs.google.com/document/d/1sKD1Lt3h4G9vzitbgoOwNX9uOrLCjdzw4bJnUIO4ixU/edit#bookmark=id.909bt66x3owh

**Comment by stanconia** [comment]
Thanks.

I don't have access to write my name but please add, Stanley Opara

**Pull request merged**
